### PR TITLE
[js/web] BiasSplitGelu and BiasAdd kernels

### PR DESCRIPTION
**Pull request by dakenf**
### Description
Two contrib kernels that supposed to speed-up StableDiffusion according to this doc https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/models/stable_diffusion/README.md

However, there is no noticable effect in speed or memory consumption. So i guess the only way to make it faster is to implement MultiHeadAttention but i'm not capable of doing that right now. So i'll focus on existing PRs and finding the JSEP kernel that produces incorrect results. It should be one of the old ones (i suspect Conv or ConvTranspose), as SD was not generating images correctly on webgpu since i started working on it. I hoped someone else would fix that by the time i finish with kernels/optimizations 😅

### CODE REVIEW

File: js/web/docs/webgpu-operators.md [hunk 1]
```diff
@@ -64,7 +66,7 @@ Do not modify directly.*
 | ReduceSumSquare | ai.onnx(1-10,11-12,13-17,18+) |  |
 | Relu | ai.onnx(6-12,13,14+) |  |
 | Reshape | ai.onnx(5-12,13,14+) | no GPU kernel |
-| Resize | ai.onnx(10,11-12,13-17,18,19+); com.ms.internal.nhwc(11-12,13-17,18,19+) | CoordinateTransformMode align_corners is not supported with downsampling |
+| Resize | ai.onnx(10,11-12,13-17,18,19+); com.ms.internal.nhwc(11-12,13-17,18,19+) |  |
```
[the remark this answers is not in the document]
Comment (dakenf) [reply js/web/docs/webgpu-operators.md:69]: No, i think i forgot to run `npm ci` after switching between branches

**Comment by satyajandhyala** [comment]
> > Can you share a (small) test case that demonstrates the problem?
> 
> Unfortunately no. I will make different builds with some JSEP kernels disabled to see which one causes the problem. Right now if i run it on webgpu, i get an image like this for any prompt. If i run it on CPU, everything works fine image

@dakenf It will help if you can you add steps to reproduce this problem?

File: js/web/lib/wasm/jsep/webgpu/ops/bias-split-gelu.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        const halfHiddenSize = outputShape[2];
+        const dataType = tensorTypeToWsglType(inputs[0].dataType);
+        const blockSize = halfHiddenSize / 256;
+        const executionCount = gridSize * 256;
+
+        const getShaderSource = (shaderHelper: ShaderHelper) => `
+  const M_SQRT2 = sqrt(2.0);
+  const TPB = 256;  // threads per block, copied from CUDA kernel
+
+  @group(0) @binding(0) var<storage, read> input : array<${dataType}>;
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/bias-split-gelu.ts:49]: please use `shaderHelper.declareVariables()`. let me know if you have any problems using IndicesHelper.

Comment (fs-eire) [reply js/web/lib/wasm/jsep/webgpu/ops/bias-split-gelu.ts:49]: well... actually for this shader maybe it is fine. I see in the shader no code is calculating incides, instead they are just using the flatten index (offset). then do not have to use the IndicesHelper

**Comment by dakenf** [comment]
> @dakenf It will help if you can you add steps to reproduce this problem?

To reproduce the error you'll need to make 64bit build of runtime. So it's quite complicated

I've found two issues that gave incorrect results
1. MatMul did not support broadcasting, fixed here https://github.com/microsoft/onnxruntime/pull/17191
2. Conv kernel gives incorrect results for StableDiffusion unet. If i comment it out, everything works fine but slow. Not sure where to start as it passes all tests

**Comment by fs-eire** [comment]
> > @dakenf It will help if you can you add steps to reproduce this problem?
> 
> To reproduce the error you'll need to make 64bit build of runtime. So it's quite complicated
> 
> I've found two issues that gave incorrect results
> 
> 1. MatMul did not support broadcasting, fixed here [[js/web] MatMul broadcasting #17191](https://github.com/microsoft/onnxruntime/pull/17191)
> 2. Conv kernel gives incorrect results for StableDiffusion unet. If i comment it out, everything works fine but slow. Not sure where to start as it passes all tests

could you help to dump a set of input/output sample (also the attributes) for the incorrect conv op so that we can take a look at it?

**Comment by dakenf** [comment]
> could you help to dump a set of input/output sample (also the attributes) for the incorrect conv op so that we can take a look at it?

yeah. there are 65 conv nodes so i feel it is going to be fun

**Comment by dakenf** [comment]
I could not find an easy way to dump inputs/outputs as they are very big but found a solution to fix my issues: https://github.com/microsoft/onnxruntime/pull/17219

**Comment by guschmue** [comment]
ci is nagging, run 'npm run format'

**Comment by dakenf** [comment]
> ci is nagging, run 'npm run format'

yup. will also add some tests and update the PR

**Comment by satyajandhyala** [comment]
@dakenf Can you try running with the latest code. I merged fixes to ConvTranspose.

**Comment by guschmue** [comment]
build fails - some imports have changed underneath: 
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1121559&view=logs&j=53557da5-d670-5ecd-2062-8e26dc1e9095&t=0c31510a-b692-5618-10ba-6036ba471b24

**Comment by dakenf** [comment]
> @dakenf Can you try running with the latest code. I merged fixes to ConvTranspose.

Yeah. Will don on the weekend. And this week emscripten got a release with all required changes (except 64bit threads) so i'll clean up 64bit PR

There's still a big issue with VRAM limit in chrome, it's ~16gb on windows. And StableDiffusion unet eats 10gb+, so it's not possible to fit unet and vae into VRAM. I'm trying to solve it with Attention+MultiHeadAttention ops and it went down to ~5gb but it does not give correct results. Most likely i've messed up with indices for packed weights or batched gemm/matmul. If i won't be able to fix it in a reasonable time, will make a draft PR and ask for your help

**Comment by dakenf** [comment]
> However i'm experiencing some other issues, like getting weird images completely unrelated to prompt. But it happens for both wasm and webgpu EPs

This is a bug in python fusion optimizer. It breaks the text encoder model, so replacing it with original one fixes the issue. However, it is quite weird because when i use python code to generate images with that broken model, it works fine

Maybe will investigate more and fill a bug later. Want to focus on optimizing InstanceNorm/LayerNorm kernels as InstanceNorm has 3 `for` loops from 0 to 40k in each invocation. And then finally revisit Attention kernel with a fresh look

**Comment by dakenf** [comment]
BTW, if you are struggling with a choice of next OP to implement, can it be NhwcConv?

**Comment by dakenf** [comment]
With MultiHeadAttention (without packed weights), Attention with some vec2/vec4 optimizations and in-place SoftMax, InstanceNorm vec2/vec4 i've narrowed unet to 5.6gb VRAM and ~2.5sec for one step

Will apply same optimizations to LayerNorm/SkipLayerNorm and implement GroupNorm to see if it will speed it up. Maybe with GroupNorm VAE will use less than 7.5gb of VRAM

So you can expect a few more PRs next week with all these stuff

https://github.com/microsoft/onnxruntime/assets/7466130/037c1ebe-7a03-4032-bcac-4e93c16a0cfd

**Comment by dakenf** [comment]
@fs-eire finally got shader-f16 extension working with latest chrome and `--enable-dawn-features=allow_unsafe_apis`

Since almost all OPs use indices helper, it will be an easy change (however i've seen some hardcoded `var x = fp32(0)`)
If that would give 2x boost and if i'll manage to implement flash attention with packed weights, it all might go down from 2.5sec to less than 1. Feels like christmas, will do tests tomorrow
image

**Comment by fs-eire** [comment]
The part of 2 operators are good. could you please revert the changes for adding test support for several other EPs? This change is helpful but should be separated.

please mention me after this change, I will try to kick the CI asap.

**Comment by dakenf** [comment]
@fs-eire i've reverted test runner changes and added fp16 support

**Review by fs-eire — APPROVED** [review]

**Review by satyajandhyala — APPROVED** [review]

**Review by guschmue — APPROVED** [review]

**Pull request merged**